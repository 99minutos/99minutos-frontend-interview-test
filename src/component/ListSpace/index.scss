.list-container{
    max-height: 600px;
    overflow-y: auto;
}
ul{
    list-style: none;
    margin: 0;
    padding: 0;
}
.text-center{
    text-align: center;
}