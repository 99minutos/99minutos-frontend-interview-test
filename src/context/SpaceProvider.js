import { createContext, useEffect, useState } from "react";
import _ from 'lodash';

const initialValue = {
    data:[],
    selected:{},
} 

export const SpaceContext = createContext();

const SpaceProvider = ({ data, children }) => {

    const [spaceData, setSpaceData] = useState(initialValue);
    
    const setData = (data) => {
        setSpaceData(prev => ({...prev, data}));
    }

    const setSelected = (id) =>{
        let selected = _.find(spaceData.data, {id});
        setSpaceData(prev => ({...prev, selected}));
    };

    useEffect(()=>{
        setSpaceData({...initialValue, data: [...data]});
    },[data])

    
    const value = { spaceData, setData, setSelected};
    
    return (
        <SpaceContext.Provider value={value}>
            {children}
        </SpaceContext.Provider>
    );
}

export default SpaceProvider;