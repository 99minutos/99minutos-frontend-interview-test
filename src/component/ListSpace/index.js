import React, { useContext } from 'react';
import { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import CardLoading from '../Card/CardLoading';
import './index.scss'
import { SpaceContext } from '../../context/SpaceProvider';
import { CircularProgress } from '@material-ui/core';
import { Redirect, useHistory } from 'react-router-dom';

const CardSpace = lazy(()=> import('../Card'));

const ListSpace = ({title, loading}) => {
    
    const { spaceData, setSelected } = useContext(SpaceContext);
    const history = useHistory();

    const handleClick = (id) => {
        setSelected(id);
        if(window.screen.width <= 768){
            history.push(`/launches/${id}`);
        }
    }

    return (
        <div className="list-container">
            <div className="text-center">
                <h1>{title}</h1>
            </div>
            <ul role="listitem">
            {
                spaceData.data.length > 0 && spaceData.data.map((data, index) =>{
                    return(
                            <Suspense fallback={<CardLoading/>} key={data.id}>
                                <CardSpace space={data} onClick={()=>handleClick(data.id)} />
                            </Suspense>
                    )
                })
            }
            </ul>
        </div>
    )
}

ListSpace.propTypes = {
    title: PropTypes.string.isRequired
}

export default ListSpace;