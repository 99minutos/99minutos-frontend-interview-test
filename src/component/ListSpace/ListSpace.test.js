import { render, screen } from "@testing-library/react";
import ListSpace from ".";
import SpaceProvider from "../../context/SpaceProvider";

describe('List Space', () => {
    const title = 'Last Launches';
    let List = null;
    const items = [
          {
              id:1,
              mission_name: "Starlink-15 (v1.0)",
              launch_date_local: "2020-10-24T11:31:00-04:00",
              links: {
                flickr_images: []
              },
              details: null,
          },
      ]

      beforeEach(()=> {List = render(<SpaceProvider data={items}>
        <ListSpace title={title}/>
    </SpaceProvider>)});

    it('should render component with title', ()=>{
        
        expect(screen.getByText(/Last Launches/i)).toBeInTheDocument();
    });

    it('should render a list of items', ()=>{
        const renderItems = screen.getAllByRole("listitem");
        expect(renderItems).toMatchSnapshot();
        expect(renderItems.length).toEqual(items.length);
        
    })
})