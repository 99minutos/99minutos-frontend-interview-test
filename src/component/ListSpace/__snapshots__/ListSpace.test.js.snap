// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Space should render a list of items 1`] = `
Array [
  <ul
    role="listitem"
  >
    <div
      class="MuiPaper-root MuiCard-root card MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardMedia-root img-container"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular makeStyles-large-1"
        >
          <img
            alt="Starlink-15 (v1.0)"
            class="MuiAvatar-img"
            src="spacecraft.png"
          />
        </div>
      </div>
      <div
        class="card-details"
      >
        <div
          class="MuiCardContent-root"
        >
          <h5
            class="title"
          >
            Starlink-15 (v1.0)
          </h5>
          <p
            class="description"
          >
            No Data Available
          </p>
          <p
            class="date"
          >
            10/24/2020
          </p>
        </div>
      </div>
    </div>
  </ul>,
]
`;
